File: ios/Runner/WebCaster.swift
import Foundation
import Network
import CommonCrypto

class WebCaster {
    static let shared = WebCaster()
    
    private var listener: NWListener?
    private var activeConnections: [NWConnection] = []
    private var isRunning = false
    private let streamingQueue = DispatchQueue(label: "dev.seven_cgpalabs.mojosnap.streaming", qos: .userInteractive)
    
    private let htmlPayload = """
    <!DOCTYPE html>
    <html>
    <head>
      <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
      <title>Mojo Snap Console</title>
      <style>
        body { background: black; margin: 0; display: flex; align-items: center; justify-content: center; height: 100vh; overflow: hidden; }
        canvas { max-width: 100%; max-height: 100%; object-fit: contain; image-rendering: pixelated; }
        #overlay { position: absolute; color: white; font-family: monospace; }
      </style>
    </head>
    <body>
      <div id="overlay">Connecting...</div>
      <canvas id="display"></canvas>
      <script>
        const canvas = document.getElementById('display');
        const ctx = canvas.getContext('2d', { alpha: false });
        const overlay = document.getElementById('overlay');
        let ws;

        function connect() {
          ws = new WebSocket('ws://' + location.host + '/ws');
          ws.binaryType = 'arraybuffer';
          
          ws.onopen = () => { overlay.style.display = 'none'; };
          ws.onclose = () => { overlay.style.display = 'block'; overlay.innerText = 'Disconnected. Reconnecting...'; setTimeout(connect, 2000); };
          
          let imgData = null;
          let buf8 = null;
          let buf32 = null;

          ws.onmessage = (e) => {
            const data = new DataView(e.data);
            const w = data.getUint16(0, true);
            const h = data.getUint16(2, true);
            
            if (canvas.width !== w || canvas.height !== h) {
              canvas.width = w;
              canvas.height = h;
              imgData = ctx.createImageData(w, h);
              buf8 = new Uint8ClampedArray(imgData.data.buffer);
              buf32 = new Uint32Array(imgData.data.buffer);
            }

            const pixels16 = new Uint16Array(e.data, 4);
            for (let i = 0; i < pixels16.length; i++) {
              const p = pixels16[i];
              const r = ((p >> 11) & 0x1F) << 3;
              const g = ((p >> 5) & 0x3F) << 2;
              const b = (p & 0x1F) << 3;
              buf32[i] = (255 << 24) | (b << 16) | (g << 8) | r;
            }
            
            imgData.data.set(buf8);
            ctx.putImageData(imgData, 0, 0);
          };
        }
        connect();
      </script>
    </body>
    </html>
    """
    
    func startServer() -> String {
        if isRunning { return currentIPAddress() }
        
        do {
            let parameters = NWParameters.tcp
            listener = try NWListener(using: parameters, on: 8080)
            
            listener?.newConnectionHandler = { [weak self] connection in
                self?.handleNewConnection(connection)
            }
            
            listener?.start(queue: .global())
            isRunning = true
            set_web_streaming(true)
            
            startStreamingLoop()
            
            return currentIPAddress()
        } catch {
            print("Failed to start WebCaster listener: \(error)")
            return ""
        }
    }
    
    func stopServer() {
        isRunning = false
        set_web_streaming(false)
        listener?.cancel()
        listener = nil
        for conn in activeConnections {
            conn.cancel()
        }
        activeConnections.removeAll()
    }
    
    private func handleNewConnection(_ connection: NWConnection) {
        connection.start(queue: .global())
        
        connection.receive(minimumIncompleteLength: 1, maximumLength: 4096) { [weak self] data, context, isComplete, error in
            guard let self = self, let data = data, let request = String(data: data, encoding: .utf8) else {
                connection.cancel()
                return
            }
            
            if request.contains("Upgrade: websocket") || request.contains("Upgrade: WebSocket") {
                self.handleWebSocketUpgrade(request: request, connection: connection)
            } else {
                self.serveHTML(connection: connection)
            }
        }
    }
    
    private func serveHTML(connection: NWConnection) {
        let response = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nConnection: close\r\n\r\n\(htmlPayload)"
        let data = response.data(using: .utf8)!
        connection.send(content: data, completion: .contentProcessed { _ in
            connection.cancel()
        })
    }
    
    private func handleWebSocketUpgrade(request: String, connection: NWConnection) {
        let lines = request.components(separatedBy: .newlines)
        var wsKey = ""
        for line in lines {
            if line.lowercased().starts(with: "sec-websocket-key:") {
                wsKey = line.components(separatedBy: ":")[1].trimmingCharacters(in: .whitespaces)
                break
            }
        }
        
        guard !wsKey.isEmpty else {
            connection.cancel()
            return
        }
        
        let magic = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
        let concatenated = wsKey + magic
        guard let data = concatenated.data(using: .utf8) else {
            connection.cancel()
            return
        }
        
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
        data.withUnsafeBytes {
            _ = CC_SHA1($0.baseAddress, CC_LONG(data.count), &digest)
        }
        
        let acceptHash = Data(digest).base64EncodedString()
        
        let response = "HTTP/1.1 101 Switching Protocols\r\n" +
                       "Upgrade: websocket\r\n" +
                       "Connection: Upgrade\r\n" +
                       "Sec-WebSocket-Accept: \(acceptHash)\r\n\r\n"
        
        connection.send(content: response.data(using: .utf8)!, completion: .contentProcessed { [weak self] error in
            if error == nil {
                self?.activeConnections.append(connection)
            } else {
                connection.cancel()
            }
        })
    }
    
    private func startStreamingLoop() {
        streamingQueue.async { [weak self] in
            guard let self = self else { return }
            
            while self.isRunning {
                let width = Int(get_web_width())
                let height = Int(get_web_height())
                
                if width > 0 && height > 0, let buffer = get_web_buffer() {
                    let pixelCount = width * height
                    let byteCount = pixelCount * 2
                    
                    var payload = Data()
                    
                    // FIN=1, OPCODE=2
                    payload.append(130)
                    
                    let payloadSize = byteCount + 4
                    if payloadSize <= 125 {
                        payload.append(UInt8(payloadSize))
                    } else if payloadSize <= 65535 {
                        payload.append(126)
                        payload.append(UInt8((payloadSize >> 8) & 0xFF))
                        payload.append(UInt8(payloadSize & 0xFF))
                    } else {
                        payload.append(127)
                        payload.append(contentsOf: [0, 0, 0, 0])
                        payload.append(UInt8((payloadSize >> 24) & 0xFF))
                        payload.append(UInt8((payloadSize >> 16) & 0xFF))
                        payload.append(UInt8((payloadSize >> 8) & 0xFF))
                        payload.append(UInt8(payloadSize & 0xFF))
                    }
                    
                    // Width and Height (little endian)
                    payload.append(UInt8(width & 0xFF))
                    payload.append(UInt8((width >> 8) & 0xFF))
                    payload.append(UInt8(height & 0xFF))
                    payload.append(UInt8((height >> 8) & 0xFF))
                    
                    // Raw pixels
                    let pixelData = Data(bytes: buffer, count: byteCount)
                    payload.append(pixelData)
                    
                    // Send to all active connections
                    let active = self.activeConnections
                    self.activeConnections = []
                    
                    for conn in active {
                        if conn.state == .ready {
                            conn.send(content: payload, completion: .contentProcessed { error in
                                if error != nil {
                                    conn.cancel()
                                }
                            })
                            self.activeConnections.append(conn)
                        }
                    }
                }
                
                usleep(16000) // ~60 FPS
            }
        }
    }
    
    private func currentIPAddress() -> String {
        var address = "127.0.0.1"
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        if getifaddrs(&ifaddr) == 0 {
            var ptr = ifaddr
            while ptr != nil {
                defer { ptr = ptr?.pointee.ifa_next }
                
                guard let interface = ptr?.pointee else { continue }
                let addrFamily = interface.ifa_addr.pointee.sa_family
                if addrFamily == UInt8(AF_INET) {
                    let name = String(cString: interface.ifa_name)
                    if name == "en0" || name == "pdp_ip0" {
                        var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                        getnameinfo(interface.ifa_addr, socklen_t(interface.ifa_addr.pointee.sa_len),
                                    &hostname, socklen_t(hostname.count),
                                    nil, socklen_t(0), NI_NUMERICHOST)
                        address = String(cString: hostname)
                    }
                }
            }
            freeifaddrs(ifaddr)
        }
        return "http://\(address):8080"
    }
}

File: android/app/src/main/kotlin/dev/seven_cgpalabs/mojosnap/WebCaster.kt
package dev.seven_cgpalabs.mojosnap

import android.util.Base64
import android.util.Log
import java.io.InputStream
import java.io.OutputStream
import java.io.ByteArrayOutputStream
import java.net.Inet4Address
import java.net.NetworkInterface
import java.net.ServerSocket
import java.net.Socket
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.zip.Deflater
import java.util.zip.DeflaterOutputStream
import kotlin.concurrent.thread

object WebCaster {
    private const val TAG = "WebCaster"
    private var serverSocket: ServerSocket? = null
    private var isRunning = false
    private var currentIp = ""

    @JvmStatic external fun setWebStreaming(streaming: Boolean)
    @JvmStatic external fun getFrameBuffer(): ByteBuffer
    @JvmStatic external fun getFrameDimensions(): IntArray
    @JvmStatic external fun getRleBuffer(): ByteBuffer
    @JvmStatic external fun getRleSize(): Int
    @JvmStatic external fun waitForNextFrame()
    @JvmStatic external fun getAudioBuffer(): ByteBuffer
    @JvmStatic external fun consumeAudioBuffer(): Int

    init {
        try {
            System.loadLibrary("native_render")
        } catch (e: UnsatisfiedLinkError) {
            Log.e(TAG, "Failed to load native library", e)
        }
    }

    fun startServer(): String {
        if (serverSocket != null) return currentIp

        try {
            val interfaces = NetworkInterface.getNetworkInterfaces()
            var ip = "127.0.0.1"
            for (intf in interfaces) {
                if (intf.name.contains("wlan") || intf.name.contains("en")) {
                    for (addr in intf.inetAddresses) {
                        if (!addr.isLoopbackAddress && addr is Inet4Address) {
                            ip = addr.hostAddress ?: ip
                        }
                    }
                }
            }
            currentIp = "http://$ip:8080"
            serverSocket = ServerSocket(8080)
            isRunning = true

            thread {
                Log.d(TAG, "Server listening on port 8080")
                while (isRunning) {
                    try {
                        val client = serverSocket?.accept() ?: break
                        thread { handleClient(client) }
                    } catch (e: Exception) {
                        if (isRunning) Log.e(TAG, "Accept error", e)
                    }
                }
            }
            return currentIp
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start server", e)
            return ""
        }
    }

    fun stopServer() {
        isRunning = false
        setWebStreaming(false)
        try {
            serverSocket?.close()
        } catch (e: Exception) { }
        serverSocket = null
    }

    private fun handleClient(socket: Socket) {
        try {
            val input = socket.getInputStream()
            val output = socket.getOutputStream()
            val reqBytes = ByteArray(4096)
            val read = input.read(reqBytes)
            if (read <= 0) {
                socket.close()
                return
            }
            
            val request = String(reqBytes, 0, read)
            if (request.contains("Upgrade: websocket", ignoreCase = true)) {
                // Handle WebSocket upgrade
                val keyRegex = "Sec-WebSocket-Key: (.*)".toRegex(RegexOption.IGNORE_CASE)
                val match = keyRegex.find(request)
                if (match != null) {
                    val key = match.groupValues[1].trim()
                    val magic = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
                    val accept = Base64.encodeToString(MessageDigest.getInstance("SHA-1").digest((key + magic).toByteArray()), Base64.NO_WRAP)

                    var response = "HTTP/1.1 101 Switching Protocols\r\n" +
                                   "Upgrade: websocket\r\n" +
                                   "Connection: Upgrade\r\n" +
                                   "Sec-WebSocket-Accept: $accept\r\n\r\n"
                    
                    output.write(response.toByteArray())
                    output.flush()

                    Log.d(TAG, "WebSocket client connected!")
                    streamFrames(socket, output)
                } else {
                    socket.close()
                }
            } else {
                // Serve HTML payload
                val response = "HTTP/1.1 200 OK\r\n" +
                               "Content-Type: text/html\r\n" +
                               "Connection: close\r\n\r\n" +
                               htmlPayload
                output.write(response.toByteArray())
                output.flush()
                socket.close()
            }
        } catch (e: java.net.SocketException) {
            // Silently ignore client disconnects
        } catch (e: Exception) {
            Log.e(TAG, "Client handler error", e)
        }
    }

    private fun streamFrames(socket: Socket, output: OutputStream) {
        val rleBuffer = getRleBuffer()
        val audioBuffer = getAudioBuffer()
        setWebStreaming(true)
        
        var frameCount = 0
        try {
            while (isRunning && !socket.isClosed) {
                // 1. Send Audio Data
                val aSize = consumeAudioBuffer()
                if (aSize > 0) {
                    val audioBytes = ByteArray(aSize)
                    audioBuffer.position(0)
                    audioBuffer.get(audioBytes, 0, aSize)
                    
                    // Header: type = 1 (audio)
                    output.write(130) // binary frame
                    
                    val payloadSize = aSize + 2 // 1 byte type, 1 byte pad
                    if (payloadSize <= 125) {
                        output.write(payloadSize)
                    } else if (payloadSize <= 65535) {
                        output.write(126)
                        output.write((payloadSize shr 8) and 0xFF)
                        output.write(payloadSize and 0xFF)
                    } else {
                        output.write(127)
                        for (i in 0..3) output.write(0)
                        output.write((payloadSize shr 24) and 0xFF)
                        output.write((payloadSize shr 16) and 0xFF)
                        output.write((payloadSize shr 8) and 0xFF)
                        output.write(payloadSize and 0xFF)
                    }
                    output.write(1) // Audio packet type
                    output.write(0) // Pad for 16-bit alignment in JS
                    output.write(audioBytes)
                    output.flush()
                }

                // 2. Send Video Data (Full 60 FPS Native HW Z-RLE)
                val dims = getFrameDimensions()
                val width = dims[0]
                val height = dims[1]
                val rleSize = getRleSize()
                    
                    if (width > 0 && height > 0 && rleSize > 0) {
                        rleBuffer.position(0)
                        val finalPayload = ByteArray(rleSize)
                        rleBuffer.get(finalPayload, 0, rleSize)
                        
                        output.write(130)
                        
                        val payloadSize = finalPayload.size + 6
                        if (payloadSize <= 125) {
                            output.write(payloadSize)
                        } else if (payloadSize <= 65535) {
                            output.write(126)
                            output.write((payloadSize shr 8) and 0xFF)
                            output.write(payloadSize and 0xFF)
                        } else {
                            output.write(127)
                            for (j in 0..3) output.write(0)
                            output.write((payloadSize shr 24) and 0xFF)
                            output.write((payloadSize shr 16) and 0xFF)
                            output.write((payloadSize shr 8) and 0xFF)
                            output.write(payloadSize and 0xFF)
                        }
                        
                        output.write(0) // Video packet type
                        output.write(0) // Padding
                        
                        val header = ByteArray(4)
                        header[0] = (width and 0xFF).toByte()
                        header[1] = ((width shr 8) and 0xFF).toByte()
                        header[2] = (height and 0xFF).toByte()
                        header[3] = ((height shr 8) and 0xFF).toByte()
                        output.write(header)
                        output.write(finalPayload)
                        output.flush()
                    }
                
                waitForNextFrame()
            }
        } catch (e: Exception) {
            println("WebCaster Client Disconnected: ${e.message}")
        } finally {
            setWebStreaming(false)
            try { socket.close() } catch (e: Exception) {}
        }
    }

    private const val htmlPayload = """
<!DOCTYPE html>
<html>
<head>
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
  <title>Mojo Snap Console</title>
  <style>
    body { background: #070714; margin: 0; display: flex; align-items: center; justify-content: center; height: 100vh; overflow: hidden; }
    canvas { display: block; }
    #overlay { position: absolute; color: white; font-family: monospace; z-index: 10; cursor: pointer; background: rgba(0,0,0,0.8); padding: 20px; border-radius: 8px;}
  </style>
</head>
<body>
  <div id="overlay">Click to Connect & Unmute Audio</div>
  <canvas id="display"></canvas>
  <script>
    const canvas = document.getElementById('display');
    const gl = canvas.getContext('webgl2', { alpha: false, antialias: false, depth: false });
    const overlay = document.getElementById('overlay');
    let ws;
    let audioCtx;
    let nextAudioTime = 0;
    
    // WebGL Shaders
    const vsSource = `#version 300 es
      in vec2 aPosition;
      in vec2 aTexCoord;
      uniform vec2 uScale;
      out vec2 vTexCoord;
      void main() {
        gl_Position = vec4(aPosition * uScale, 0.0, 1.0);
        vTexCoord = aTexCoord;
      }
    `;
    const fsSource = `#version 300 es
      precision mediump float;
      in vec2 vTexCoord;
      uniform sampler2D uTexture;
      out vec4 fragColor;
      void main() {
        vec4 color = texture(uTexture, vTexCoord);
        // Subtle CRT scanline effect
        float scanline = sin(vTexCoord.y * 800.0) * 0.08;
        color.rgb -= scanline;
        fragColor = vec4(color.rgb, 1.0);
      }
    `;
    
    function compileShader(type, source) {
      const shader = gl.createShader(type);
      gl.shaderSource(shader, source);
      gl.compileShader(shader);
      return shader;
    }
    
    const program = gl.createProgram();
    gl.attachShader(program, compileShader(gl.VERTEX_SHADER, vsSource));
    gl.attachShader(program, compileShader(gl.FRAGMENT_SHADER, fsSource));
    gl.linkProgram(program);
    gl.useProgram(program);
    
    // Quad Data
    const vertices = new Float32Array([
      -1, -1,  0, 1,
       1, -1,  1, 1,
      -1,  1,  0, 0,
       1,  1,  1, 0,
    ]);
    const vbo = gl.createBuffer();
    gl.bindBuffer(gl.ARRAY_BUFFER, vbo);
    gl.bufferData(gl.ARRAY_BUFFER, vertices, gl.STATIC_DRAW);
    
    const aPosition = gl.getAttribLocation(program, 'aPosition');
    const aTexCoord = gl.getAttribLocation(program, 'aTexCoord');
    const uScale = gl.getUniformLocation(program, 'uScale');
    
    gl.enableVertexAttribArray(aPosition);
    gl.vertexAttribPointer(aPosition, 2, gl.FLOAT, false, 16, 0);
    gl.enableVertexAttribArray(aTexCoord);
    gl.vertexAttribPointer(aTexCoord, 2, gl.FLOAT, false, 16, 8);
    
    // Texture
    const texture = gl.createTexture();
    gl.bindTexture(gl.TEXTURE_2D, texture);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
    
    let currentWidth = 0;
    let currentHeight = 0;
    
    function updateViewport() {
      if (currentWidth === 0) return;
      canvas.width = window.innerWidth;
      canvas.height = window.innerHeight;
      gl.viewport(0, 0, canvas.width, canvas.height);
      
      const coreAspect = currentWidth / currentHeight;
      const winAspect = canvas.width / canvas.height;
      
      let scaleX = 1.0;
      let scaleY = 1.0;
      
      if (winAspect > coreAspect) {
        scaleX = coreAspect / winAspect;
      } else {
        scaleY = winAspect / coreAspect;
      }
      gl.uniform2f(uScale, scaleX, scaleY);
    }
    
    window.addEventListener('resize', updateViewport);

    overlay.onclick = () => {
      audioCtx = new (window.AudioContext || window.webkitAudioContext)({ sampleRate: 44100 });
      connect();
    };

    function connect() {
      ws = new WebSocket('ws://' + location.host + '/ws');
      ws.binaryType = 'arraybuffer';
      
      ws.onopen = () => { overlay.style.display = 'none'; };
      ws.onclose = () => { overlay.style.display = 'block'; overlay.innerText = 'Disconnected. Click to Reconnect.'; ws = null; };
      
      let imgData = null;
      let buf8 = null;
      let buf32 = null;

      ws.onmessage = (e) => {
        const data = new DataView(e.data);
        const type = data.getUint8(0);
        
        if (type === 1) { // Audio
            if (!audioCtx) return;
            const int16Data = new Int16Array(e.data, 2);
            const samples = int16Data.length / 2;
            const buffer = audioCtx.createBuffer(2, samples, 44100);
            const left = buffer.getChannelData(0);
            const right = buffer.getChannelData(1);
            for(let i=0; i<samples; i++) {
                left[i] = int16Data[i*2] / 32768.0;
                right[i] = int16Data[i*2+1] / 32768.0;
            }
            const source = audioCtx.createBufferSource();
            source.buffer = buffer;
            source.connect(audioCtx.destination);
            
            // Anti-Drift: Snap to +50ms if we underrun OR if we drift too far into the future!
            if (nextAudioTime < audioCtx.currentTime || nextAudioTime > audioCtx.currentTime + 0.1) {
                nextAudioTime = audioCtx.currentTime + 0.05;
            }
            
            source.start(nextAudioTime);
            nextAudioTime += buffer.duration;
            return;
        }

        if (type === 0) { // Video
            const w = data.getUint16(2, true);
            const h = data.getUint16(4, true);
            
            if (w === 0 || h === 0) return;
            
            if (currentWidth !== w || currentHeight !== h) {
              currentWidth = w;
              currentHeight = h;
              updateViewport();
            }

            const rleBytes = new Uint8Array(e.data, 6);
            const pixels16 = new Uint16Array(w * h);
            let readIdx = 0;
            let writeIdx = 0;
            
            while (readIdx < rleBytes.length && writeIdx < pixels16.length) {
                const c = rleBytes[readIdx++];
                if (c < 128) {
                    const count = c + 1;
                    for (let i = 0; i < count; i++) {
                        pixels16[writeIdx++] = rleBytes[readIdx] | (rleBytes[readIdx+1] << 8);
                        readIdx += 2;
                    }
                } else {
                    const count = (c - 128) + 2;
                    const color = rleBytes[readIdx] | (rleBytes[readIdx+1] << 8);
                    readIdx += 2;
                    for (let i = 0; i < count; i++) {
                        pixels16[writeIdx++] = color;
                    }
                }
            }
            gl.bindTexture(gl.TEXTURE_2D, texture);
            gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGB565, w, h, 0, gl.RGB, gl.UNSIGNED_SHORT_5_6_5, pixels16);
            gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
        }
      };
    }
  </script>
</body>
</html>
"""
}
